crates/nomination/src/types.rs

```diff
@@ -0,0 +1,533 @@
+    }
+
+    pub fn slash_nominators_by(&mut self, amount_to_slash: DOT<T>) -> DispatchResult {
+        let vault_clone = self.data.clone();
+        for (nominator_id, _) in &vault_clone.nominators {
+            let nominated_collateral_to_slash =
+                self.scale_amount_by_nominator_proportion(amount_to_slash, nominator_id.clone())?;
+            self.decrease_nominator_collateral(nominator_id.clone(), nominated_collateral_to_slash)?;
+        }
+        Ok(())
+    }
+
+    pub fn scale_amount_by_nominator_proportion(
+        &mut self,
+        amount: DOT<T>,
+        nominator_id: T::AccountId,
+    ) -> Result<DOT<T>, DispatchError> {
+        let amount_u128 = Module::<T>::dot_to_u128(amount)?;
+        let nominated_collateral_proportion = self.get_nominator_collateral_proportion(nominator_id.clone())?;
+        let nominated_collateral_to_slash_u128 = nominated_collateral_proportion
+            .checked_mul(amount_u128)
+            .ok_or(Error::<T>::ArithmeticOverflow)?;
+        Module::<T>::u128_to_dot(nominated_collateral_to_slash_u128)
+    }
+
+    pub fn get_nominator_collateral_proportion(&mut self, nominator_id: T::AccountId) -> Result<u128, DispatchError> {
+        let nominator = self
+            .data
+            .nominators
+            .get(&(nominator_id.clone()))
+            .ok_or(Error::<T>::NominatorNotFound)?;
+        let nominated_collateral_u128 = Module::<T>::dot_to_u128(nominator.collateral)?;
+        let total_nominated_collateral_u128 = Module::<T>::dot_to_u128(self.data.total_nominated_collateral)?;
+        Ok(nominated_collateral_u128
+            .checked_div(total_nominated_collateral_u128)
+            .ok_or(Error::<T>::ArithmeticUnderflow)?)
+    }
+
+    // Operator functionality
+
+    pub fn add_pending_operator_withdrawal(
+        &mut self,
+        request_id: H256,
+        collateral_to_withdraw: DOT<T>,
+        maturity: T::BlockNumber,
+    ) -> DispatchResult {
+        // If `collateral_to_withdraw` is larger than (operator_collateral - collateral_to_be_withdrawn),
+        // the following throws an error.
+        let operator_collateral = self.get_operator_collateral()?;
+        let remaining_operator_collateral = operator_collateral
+            .checked_sub(&self.data.collateral_to_be_withdrawn)
+            .ok_or(Error::<T>::ArithmeticUnderflow)?
+            .checked_sub(&collateral_to_withdraw)
+            .ok_or(Error::<T>::ArithmeticUnderflow)?;
+
+        // Trigger forced refunds if remaining nominated collateral would
+        // exceed the Max Nomination Ratio.
+        let max_nominatable_collateral = self.get_max_nominatable_collateral(remaining_operator_collateral)?;
+        let nominated_collateral_to_force_refund = self
+            .data
+            .total_nominated_collateral
+            .saturating_sub(max_nominatable_collateral);
+        let zero_dot: DOT<T> = 0u32.into();
+        if !nominated_collateral_to_force_refund.eq(&zero_dot) {
```

Comment:
Use `nominated_collateral_to_force_refund.is_zero()`. Also, should we return an error if we were unable to withdraw anything?

Reply:
We shouldn't throw an error if `nominated_collateral_to_force_refund` is zero. This conditional checks if, besides adding an Operator withdrawal request, we also need to force refund the Nominators. It will be true only when the Operator's withdrawal would cause the Max Nomination Ratio to be exceeded. Neither of the two cases handled by this conditional represents an error